nifi-extension-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/CaptureBoxEvents.java

```diff
@@ -0,0 +1,270 @@
+                    Setting this value higher allows more messages to be buffered in memory during surges of incoming messages, but increases the total
+                    memory used by the processor during these surges.
+                    """)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("10000")
+            .required(true)
+            .build();
+
+    private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = List.of(
+            BoxClientService.BOX_CLIENT_SERVICE,
+            MAX_MESSAGE_QUEUE_SIZE
+    );
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("Events received successfully will be sent out this relationship.")
+            .build();
+
+    private static final Set<Relationship> RELATIONSHIPS = Set.of(REL_SUCCESS);
+
+    private volatile BoxAPIConnection boxAPIConnection;
+    private volatile EventStream eventStream;
+    protected volatile LinkedBlockingQueue<BoxEvent> events;
+    private volatile AtomicLong position = new AtomicLong(0);
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTY_DESCRIPTORS;
+    }
+
+    @Override
+    public final Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) {
+        final BoxClientService boxClientService = context.getProperty(BoxClientService.BOX_CLIENT_SERVICE).asControllerService(BoxClientService.class);
+        boxAPIConnection = boxClientService.getBoxApiConnection();
+
+        try {
+            final String position = context.getStateManager().getState(Scope.CLUSTER).get(POSITION_KEY);
+            if (position != null) {
+                // we resume from the last known position
+                eventStream = new EventStream(boxAPIConnection, Long.parseLong(position));
+            } else {
+                eventStream = new EventStream(boxAPIConnection);
+            }
+        } catch (Exception e) {
+            throw new ProcessException("Could not retrieve processor state", e);
+        }
+
+        final int eventsCapacity = context.getProperty(MAX_MESSAGE_QUEUE_SIZE).asInteger();
+        if (events == null) {
+            events = new LinkedBlockingQueue<>(eventsCapacity);
+        } else {
+            // create new one with events from the old queue in case capacity has changed
+            final LinkedBlockingQueue<BoxEvent> newQueue = new LinkedBlockingQueue<>(eventsCapacity);
+            newQueue.addAll(events);
+            events = newQueue;
+        }
+
+        eventStream.addListener(new EventListener() {
+
+            @Override
+            public void onEvent(BoxEvent event) {
+                if (!events.offer(event)) {
```

Comment:
Reviewing the `EventStream` behavior, it seems like a better option would be to use `put()`, which would block the EventStream Thread until the events queue has capacity. This would avoid errors, and dropping data.